File: QuasarCode_Library_Games/Dice/DiceCup.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Linq;

using QuasarCode.Library.Tools;

namespace QuasarCode.Library.Games.Dice
{
    /// <summary>
    /// An immutable collection of NDice objects that are rolled together.
    /// </summary>
    public sealed class DiceCup : IStaticDiceCup
    {
        /// <summary>
        /// Array of the NDice objects rolled by the cup.
        /// </summary>
        public IDice[] AllDice { get; private set; }

        /// <summary>
        /// Number of dice stored.
        /// </summary>
        public int Count { get { return AllDice.Length; } }

        /// <summary>
        /// Creates a new dice cup.
        /// </summary>
        /// <param name="sides">The number of sides on each dice</param>
        /// <param name="noOfDice">Number of dice to create</param>
        public DiceCup(int sides, int noOfDice)
        {
            List<NDice> allDice = new List<NDice>();
            
            foreach (NDice dice in NDice.MultipleDice(sides, noOfDice))
            {
                allDice.Add(dice);
            }

            AllDice = allDice.ToArray();
        }

        /// <summary>
        /// Creates a new dice cup.
        /// </summary>
        /// <param name="sides">Side numbers representing each unique dice</param>
        public DiceCup(params int[] sides)
        {
            List<NDice> allDice = new List<NDice>();

            foreach (NDice dice in NDice.MultipleDice(sides))
            {
                allDice.Add(dice);
            }

            AllDice = allDice.ToArray();
        }

        /// <summary>
        /// Creates a new dice cup.
        /// </summary>
        /// <param name="sides">Collection of side numbers representing unique dice</param>
        public DiceCup(ICollection<int> sides)
        {
            List<NDice> allDice = new List<NDice>();

            foreach (NDice dice in NDice.MultipleDice(sides))
            {
                allDice.Add(dice);
            }

            AllDice = allDice.ToArray();
        }

        /// <summary>
        /// Rolls all the dice. Returns each individual result.
        /// </summary>
        /// <param name="suppressEmptyExeption">Suppress the exeption raised when the cup is empty</param>
        /// <returns>Array of integers</returns>
        public int[] Roll(bool suppressEmptyExeption = false)
        {
            if (!suppressEmptyExeption && Count == 0)
            {
                throw new InvalidOperationException("There are no dice in the cup.");
            }

            List<int> results = new List<int>();

            foreach (NDice dice in AllDice)
            {
                results.Add(dice.Roll());
            }

            return results.ToArray();
        }

        /// <summary>
        /// Rolls all the dice. Returns the sum of the results.
        /// </summary>
        /// <param name="suppressEmptyExeption">Suppress the exeption raised when the cup is empty</param>
        /// <returns>Integer</returns>
        public int RollTotal(bool suppressEmptyExeption = false)
        {
            if (!suppressEmptyExeption && Count == 0)
            {
                throw new InvalidOperationException("There are no dice in the cup.");
            }

            int result = 0;

            foreach (NDice dice in AllDice)
            {
                result += dice.Roll();
            }

            return result;
        }
    }
}

File: old-code/src/QuasarCode/Games/Dice/_DiceCup.py
from collections.abc import MutableSequence
from ._NDice import NDice

class DiceCup(MutableSequence):
    """
    Collection of NDice objects that are rolled together

    Atributes:
        list AllDice --> list of the NDice objects rolled by the cup
        int Count --> number of dice stored
        bool Dynamic --> whether or not the cup is mutable

    __init__():
        list sides --> list of side numbers representing unique dice. For multiple of the same number, use a single element and alter noOfDice
        int noOfDice --> number of dice to create (deafult is length of sides list)
        bool dynamic --> whether or not the cup is mutable
    """
    def __init__(self, sides: list, noOfDice: int = None, dynamic: bool = False):
        #? super().__init__()

        self.Dynamic = dynamic
        self.AllDice: list = NDice.MultipleDice(sides, noOfDice)

        self.Count: int = property(fget = lambda self: len(self.AllDice))

        if not dynamic and len(self.AllDice) == 0:
            raise ValueError("No dice were specified for a static dice cup.")

    def Roll(self, suppressEmptyExeption = False):
        """
        Rolls all the dice and returns each individual result
        """
        if not suppressEmptyExeption and self.Count == 0:
            raise ValueError("There are no dice in the cup.");

        results = []

        for dice in self.AllDice:
            results.append(dice.Roll())

        return results

    def RollTotal(self, suppressEmptyExeption = False):
        """
        Rolls all the dice and returns the sum of the results
        """
        return sum(self.Roll(suppressEmptyExeption))

    def __getitem__(self, key):
        if not self.Dynamic:
            raise RuntimeError("Invalid operation - the dice cup is not dynamic!")

        return self.AllDice.__getitem__(key)

    def __setitem__(self, key, value):
        if not self.Dynamic:
            raise RuntimeError("Invalid operation - the dice cup is not dynamic!")

        self.AllDice.__setitem__(key, value)

    def __delitem__(self, key):
        if not self.Dynamic:
            raise RuntimeError("Invalid operation - the dice cup is not dynamic!")

        self.AllDice.__delitem__(key)

    def __len__(self):
        return len(self.AllDice)

    def insert(self, index, object):
        if not self.Dynamic:
            raise RuntimeError("Invalid operation - the dice cup is not dynamic!")

        self.AllDice.insert(index, object)

    def __iadd__(self, value):
        self.append(value)
        return self